lib/Dialect/Moore/Transforms/MergeAssignments.cpp

```diff
@@ -0,0 +1,79 @@
+//===- MergeAssignments.cpp - Merge declaration and assignment ------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the MergeAssignments pass.
+// Find easy uses of declaration and assignments and merge them into
+// assigned_variable. Easy use represents the declaration doesn't be performed
+// bit slice. Like bit [127:0] b; assign b = 128'b0;
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Moore/MooreOps.h"
+#include "circt/Dialect/Moore/MoorePasses.h"
+#include "mlir/IR/Dialect.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+namespace circt {
+namespace moore {
+#define GEN_PASS_DEF_MERGEASSIGNMENTS
+#include "circt/Dialect/Moore/MoorePasses.h.inc"
+} // namespace moore
+} // namespace circt
+
+using namespace circt;
+using namespace moore;
+
+namespace {
+struct MergeAssignmentsPass
+    : public circt::moore::impl::MergeAssignmentsBase<MergeAssignmentsPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+std::unique_ptr<mlir::Pass> circt::moore::createMergeAssignmentsPass() {
+  return std::make_unique<MergeAssignmentsPass>();
+}
+
+// TODO: The net can be driven multiple times. However, the related rule is
+// complicated. So we will implement this in the future.
+
+// Only collect the easy declaration and its value at module level.
+static void collectAssignmets(SVModuleOp moduleOp,
+                              DenseMap<Value, Value> &assignments) {
+  moduleOp->walk([&](Operation *op) {
+    TypeSwitch<Operation *>(op).Case<ContinuousAssignOp>([&](auto op) {
+      if (isa<VariableOp, NetOp>(op.getDst().getDefiningOp())) {
+        if (!assignments.lookup(op.getDst()))
+          assignments[op.getDst()] = op.getSrc();
+        else
+          mlir::emitError(op.getLoc())
+              << "Unsupported drive the same wire '"
+              << op.getDst().template getDefiningOp<NetOp>().getName()
+              << "' two times now";
```

Comment:
Instead of emitting an error here, could we just skip this variable/net op and not try to merge its assignment? It's legal for a net op to have multiple continuous assignments for example, and having this code here would cause the compiler to error out despite the legal use case.

The assignment merging feels like an opportunistic optimization that brings the IR into a form that's easier to analyze, but it's technically not needed for correctness. We could always keep the original var/net op and assignment op and lower them into `llhd.sig` and `llhd.drv`.

Reply:
Have substance! I'll tweak this error.

Yeah, I keep the original var/net and assignment ops according to what you meant.  However, I have a question about lowering them into `llhd.sig` and `llhd.drv`.  If we keep them, they also slow the speed of the simulation similar to directly lower into `llhd.sig` and `llhd.drv` without `assigned_variable`, right?

Reply:
Yes I think the simulation will be slower. But it will still be correct. From that point of view, merging `var+assign` into `assigned_var` is a pure optimization that doesn't affect correctness. It just makes the simulation a bit faster. Therefore it should never emit an error, because it is optional. Instead, if the merging cannot happen for whatever reason, it should just silently not happen 🙂

Reply:
Thanks for the detailed explanation.